compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1330,7 +1331,35 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) -> Ty<'tcx> {
         let rcvr_t = self.check_expr(rcvr);
         // no need to check for bot/err -- callee does that
-        let rcvr_t = self.structurally_resolve_type(rcvr.span, rcvr_t);
+        let rcvr_t = if let ExprKind::Index(_, index, _) = rcvr.kind {
+            let ty = self.try_structurally_resolve_type(rcvr.span, rcvr_t);
+
+            if !ty.is_ty_var() {
+                ty
+            } else {
+                let e = self
+                    .tainted_by_errors()
+                    .or_else(|| self.report_ambiguity_errors())
```

Comment:
I don't believe it's correct to call `report_ambiguity_errors` here. Please look at the definition of that function -- since it calls `collect_remaining_errors`, it essentially forces all obligations that are currently waiting on inference to be reported as errors. It's only meant to be called at the *end* of type checking.

Reply:
Calling `report_ambiguity_errors` cause the inference of `recv_t` has failed, so I want to check if there are any remaining errors and emit them. If no such errors, then just emit the inference failure of `recv`.

Reply:
Is there any ways to do such things? cause I'd like to suggest like what `thing[i];` did.